File: Sources/NotoCore/Models/Graph.swift
public enum GraphFilter: String, CaseIterable, Identifiable, Equatable {
    case all
    case local
    case lectureOnly
    case orphans

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .all: return "All Notes"
        case .local: return "Local Graph"
        case .lectureOnly: return "Lecture Notes"
        case .orphans: return "Orphans"
        }
    }
}

public struct GraphNode: Identifiable, Equatable {
    public let id: String
    public let title: String
    public let path: String
    public let backlinksCount: Int
    public let outgoingCount: Int

    public init(id: String, title: String, path: String, backlinksCount: Int, outgoingCount: Int) {
        self.id = id
        self.title = title
        self.path = path
        self.backlinksCount = backlinksCount
        self.outgoingCount = outgoingCount
    }

    public var degree: Int {
        backlinksCount + outgoingCount
    }
}

public struct GraphEdge: Identifiable, Equatable {
    public let id: String
    public let source: String
    public let target: String
    public let weight: Double

    public init(id: String, source: String, target: String, weight: Double) {
        self.id = id
        self.source = source
        self.target = target
        self.weight = weight
    }
}

public struct KnowledgeGraph: Equatable {
    public let nodes: [GraphNode]
    public let edges: [GraphEdge]

    public init(nodes: [GraphNode], edges: [GraphEdge]) {
        self.nodes = nodes
        self.edges = edges
    }
}

File: landing/src/noto-core/graph.ts
// Faithful port of Sources/NotoCore/Lib/GraphBuilder.swift
import type {
  GraphEdge,
  GraphFilter,
  GraphNode,
  KnowledgeGraph,
  MetadataCache,
  VaultFile,
} from "./types";

const LECTURE_PREFIX = "AI Lecture Notes/";

export function buildGraph(files: VaultFile[], cache: MetadataCache): KnowledgeGraph {
  const nodes: GraphNode[] = files.map((file) => {
    const meta = cache.filesById[file.id];
    const backlinksCount = meta?.backlinks.length ?? 0;
    const outgoingCount = meta?.outgoingLinks.length ?? 0;
    return {
      id: file.id,
      title: file.title,
      path: file.path,
      backlinksCount,
      outgoingCount,
      degree: backlinksCount + outgoingCount,
    };
  });

  const edges: GraphEdge[] = [];
  for (const file of files) {
    const meta = cache.filesById[file.id];
    if (!meta) continue;
    for (const targetTitle of meta.outgoingLinks) {
      const targetId = cache.fileIdByTitle[targetTitle];
      if (targetId === undefined) continue;
      edges.push({
        id: `${file.id}->${targetId}`,
        source: file.id,
        target: targetId,
        weight: 1,
      });
    }
  }

  return { nodes, edges };
}

export function filterGraph(
  graph: KnowledgeGraph,
  mode: GraphFilter,
  activeFileId: string,
): KnowledgeGraph {
  switch (mode) {
    case "all":
      return graph;
    case "local":
      return subgraph(graph, localNodeIds(graph, activeFileId));
    case "lectureOnly": {
      const ids = new Set(
        graph.nodes.filter((n) => n.path.startsWith(LECTURE_PREFIX)).map((n) => n.id),
      );
      return subgraph(graph, ids);
    }
    case "orphans": {
      const connected = new Set<string>();
      for (const e of graph.edges) {
        connected.add(e.source);
        connected.add(e.target);
      }
      const ids = new Set(graph.nodes.filter((n) => !connected.has(n.id)).map((n) => n.id));
      return subgraph(graph, ids);
    }
  }
}

function localNodeIds(graph: KnowledgeGraph, activeFileId: string): Set<string> {
  const ids = new Set<string>([activeFileId]);
  for (const edge of graph.edges) {
    if (edge.source === activeFileId) ids.add(edge.target);
    if (edge.target === activeFileId) ids.add(edge.source);
  }
  return ids;
}

function subgraph(graph: KnowledgeGraph, ids: Set<string>): KnowledgeGraph {
  return {
    nodes: graph.nodes.filter((n) => ids.has(n.id)),
    edges: graph.edges.filter((e) => ids.has(e.source) && ids.has(e.target)),
  };
}
